packages/tabsheet/src/vaadin-tabsheet-mixin.js

```diff
@@ -0,0 +1,183 @@
+/**
+ * @license
+ * Copyright (c) 2022 - 2023 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import { DelegateStateMixin } from '@vaadin/component-base/src/delegate-state-mixin.js';
+import { OverflowController } from '@vaadin/component-base/src/overflow-controller.js';
+import { SlotController } from '@vaadin/component-base/src/slot-controller.js';
+import { SlotObserver } from '@vaadin/component-base/src/slot-observer.js';
+import { generateUniqueId } from '@vaadin/component-base/src/unique-id-utils.js';
+
+/**
+ * @private
+ * A controller which observes the <vaadin-tabs> slotted to the tabs slot.
+ */
+class TabsSlotController extends SlotController {
+  constructor(host) {
+    super(host, 'tabs');
+    this.__tabsItemsChangedListener = this.__tabsItemsChangedListener.bind(this);
+    this.__tabsSelectedChangedListener = this.__tabsSelectedChangedListener.bind(this);
+  }
+
+  /** @private */
+  __tabsItemsChangedListener() {
+    this.host._setItems(this.tabs.items);
+  }
+
+  /** @private */
+  __tabsSelectedChangedListener() {
+    this.host.selected = this.tabs.selected;
+  }
+
+  initCustomNode(tabs) {
+    if (!(tabs instanceof customElements.get('vaadin-tabs'))) {
+      throw Error('The "tabs" slot of a <vaadin-tabsheet> must only contain a <vaadin-tabs> element!');
+    }
+    this.tabs = tabs;
+    tabs.addEventListener('items-changed', this.__tabsItemsChangedListener);
+    tabs.addEventListener('selected-changed', this.__tabsSelectedChangedListener);
+    this.host.__tabs = tabs;
+    this.host.stateTarget = tabs;
+    this.__tabsItemsChangedListener();
+  }
+
+  teardownNode(tabs) {
+    this.tabs = null;
+    tabs.removeEventListener('items-changed', this.__tabsItemsChangedListener);
+    tabs.removeEventListener('selected-changed', this.__tabsSelectedChangedListener);
+    this.host.__tabs = null;
+    this.host._setItems([]);
+    this.host.stateTarget = undefined;
+  }
+}
+
+/**
+ * @polymerMixin
+ * @mixes DelegateStateMixin
+ */
+export const TabSheetMixin = (superClass) =>
+  class extends DelegateStateMixin(superClass) {
+    static get properties() {
+      return {
+        /**
+         * The list of `<vaadin-tab>`s from which a selection can be made.
+         * It is populated from the elements passed inside the slotted
+         * `<vaadin-tabs>`, and updated dynamically when adding or removing items.
+         *
+         * Note: unlike `<vaadin-combo-box>`, this property is read-only.
+         * @type {!Array<!Tab> | undefined}
+         */
+        items: {
+          type: Array,
+          readOnly: true,
+          notify: true,
+        },
```

Comment:
nit: let's add newlines between properties here and below.

Reply:
Done